src/sentry/snuba/metrics/datasource.py

```diff
@@ -488,114 +621,57 @@ def get_series(projects: Sequence[Project], query: QueryDefinition) -> dict:
-                    # represents transaction
-                    # {
-                    #     (3, 2): [{"metric_id": 4, "project_id": 3, "tags[1]": 2, "p50": [11.0]}],
-                    #     (3, 3): [{"metric_id": 4, "project_id": 3, "tags[1]": 3, "p50": [5.0]}],
-                    # }
-                    snuba_query_data_dict = {}
-                    for data_elem in snuba_query_res["data"]:
-                        snuba_query_data_dict.setdefault(
-                            tuple(data_elem[col] for col in groupby_tags), []
-                        ).append(data_elem)
-
-                    # Order the results according to the results of the initial query, so that when
-                    # the results dict is passed on to `SnubaResultsConverter`, it comes out ordered
-                    # Ordered conditions might for example look something like this
-                    # {..., ('project_id', 'tags[1]'): [(3, 3), (3, 2)]}, then we end up with
-                    # {
-                    #     "totals": {
-                    #         "data": [
-                    #             {
-                    #               "metric_id": 5, "project_id": 3, "tags[1]": 3, "count_unique": 5
-                    #             },
-                    #             {
-                    #               "metric_id": 5, "project_id": 3, "tags[1]": 2, "count_unique": 1
-                    #             },
-                    #         ]
-                    #     }
-                    # }
-                    for group_tuple in ordered_tag_conditions[groupby_tags]:
-                        results[entity][key]["data"] += snuba_query_data_dict.get(group_tuple, [])
     else:
         snuba_queries, fields_in_entities = SnubaQueryBuilder(projects, query).get_snuba_queries()
+        group_limit_filters = None
+
         for entity, queries in snuba_queries.items():
             results.setdefault(entity, {})
             for key, snuba_query in queries.items():
-                if snuba_query is None:
-                    continue
-                results[entity][key] = raw_snql_query(
-                    snuba_query, use_cache=False, referrer=f"api.metrics.{key}"
-                )
+                if group_limit_filters:
+                    snuba_query = _apply_group_limit_filters(snuba_query, group_limit_filters)
+
+                snuba_result = raw_snql_query(
+                    snuba_query,
+                    use_cache=False,
+                    referrer=f"api.metrics.{key}",
+                )["data"]
+
+                snuba_limit = snuba_query.limit.limit if snuba_query.limit else None
+                if not group_limit_filters and snuba_limit and len(snuba_result) == snuba_limit:
+                    group_limit_filters = _get_group_limit_filters(query, snuba_result)
+                    if group_limit_filters:
+                        _prune_extra_groups(results, group_limit_filters)
+
```

Comment:
There are two scenarios that I think are not handled ideally here, related to time series:

1. The `totals` query is the first to hit the limit, let's say 200. We then add a filter of 200 groups to the time series queries, of let's 90 days. But the time series would now need to query 90 * 200 = 18000 rows, so the snuba query for time series (which is capped at 10000) would still return incomplete results.
2. More likely scenario: The `totals` query runs first, and does not hit the limit, but then the `series` query runs and does hit the limit. `_prune_empty_groups` will now remove extra groups from the `totals` result, but we could have prevented querying these groups in the first place.

I think both scenarios could be fixed by first computing the max. number of groups, as we do in the sessions api: https://github.com/getsentry/sentry/blob/92fde98b9345678d858a3b09f5ddd81a9b7bd7a9/src/sentry/snuba/sessions_v2.py#L467

Reply:
Thank you, that's a great catch! I'll fix this by adjusting limit computation in the query builder.

Reply:
With the `max_totals` computation I think both scenarios are now addressed.